### PR TITLE
Implementing skip_blank option

### PR DESCRIPTION
**Pull request by aroberge**
I've implemented the `skip_blank` option by adding a new line type, and adding a new `Option`.
Here's an example where the default (skipping blank lines) is used:
```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 13, in <module>
       7 |     length = len(seq)
       9 |     return seq[length]
      12 | seq = [1, 2, 3]
-->   13 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 9, in last_item
       5 | def last_item(seq):
       7 |     length = len(seq)
-->    9 |     return seq[length]
                      ^^^^^^^^^^^
IndexError: list index out of range
```
And here's an example where `skip_blank` is set to `False`:
```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 14, in <module>
       7 |     length = len(seq)
       : |
      10 |     return seq[length]
       : |
      13 | seq = [1, 2, 3]
-->   14 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 10, in last_item
       5 | def last_item(seq):
       6 |
       7 |     length = len(seq)
       : |
-->   10 |     return seq[length]
                      ^^^^^^^^^^^
IndexError: list index out of range
```

### CODE REVIEW

**Comment by aroberge** [comment]
Here's a different possibility when multiple lines are skipped:
```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 14, in <module>
       7 |     length = len(seq)
       :
      10 |     return seq[length]
       :
      13 | seq = [1, 2, 3]
-->   14 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 10, in last_item
       5 | def last_item(seq):
       6 |
       7 |     length = len(seq)
       :
-->   10 |     return seq[length]
                      ^^^^^^^^^^^
IndexError: list index out of range
```
I admit that it is a bit of a contrived examples, as there are rarely two blank lines in the code, except before functions and class definitions. Still, I have a slight preference for this one, using only ` : ` and not ` : | `, as it is less cluttered and shows better the gap -- I think.

File: stack_data/formatting.py [hunk 1]
```diff
@@ -189,9 +191,19 @@ def format_line(self, line: Line) -> str:
...
                 )
-
         return result
 
+
+    def format_blank_lines(self, blank_line):
+        result = ""
+        if self.current_line_indicator:
+            result = " " * len(self.current_line_indicator)
+        result = result or "   "
```
Comment (alexmojaki) [inline stack_data/formatting.py:201]:
```suggestion
        if self.current_line_indicator:
            result = " " * len(self.current_line_indicator)
        else:
            result = "   "
```

This seems equivalent to me.

File: stack_data/formatting.py [hunk 2]
```diff
@@ -138,6 +138,8 @@ def format_frame(self, frame: Union[FrameInfo, FrameType, TracebackType]) -> Ite
         for line in frame.lines:
             if isinstance(line, Line):
                 yield self.format_line(line)
+            elif isinstance(line, BlankLines) and self.show_linenos:
```
Comment (alexmojaki) [inline stack_data/formatting.py:141]: This implies that `not options.skip_blank and not self.show_linenos` may or may not fail on `assert_(line is LINE_GAP)`. If that combination shouldn't be allowed, it should be checked in `__init__`.

**Review by alexmojaki — CHANGES_REQUESTED** [review]

**Comment by alexmojaki** [comment]
> Still, I have a slight preference for this one, using only `:` and not `: |`, as it is less cluttered and shows better the gap -- I think.

You could add a `Formatter` keyword argument that requires a string which will be formatted in place and right-aligned to line up with the `|` from normal lines, so `":  "` (that's two spaces at the end) and `": |"` would both be allowed and satisfy both your ideas. Some might also want to use a proper unicode vertical ellipsis with 3 dots.

File: stack_data/core.py [hunk 3]
```diff
@@ -171,13 +171,15 @@ def __init__(
             after: int = 1,
             include_signature: bool = False,
             max_lines_per_piece: int = 6,
-            pygments_formatter=None
+            pygments_formatter=None,
+            skip_blank = True
```
[the remark this answers is not in the document]
Comment (aroberge) [reply stack_data/core.py:175]: Ok, I will see what I can do for this.

File: stack_data/core.py [hunk 4]
```diff
@@ -336,6 +347,8 @@ def range_from_node(
                     return None
         else:
             range_end = len(self.text)
+        if range_start == range_end == 0:
```
[the remark this answers is not in the document]
Comment (aroberge) [reply stack_data/core.py:350]: A blank line has zero indentation: it should not be used to calculate the common minimal indentation of a block of code. I will add a comment to that effect.

File: stack_data/core.py [hunk 5]
```diff
@@ -521,6 +534,10 @@ def __init__(
         self.code = frame.f_code
         self.options = options or Options()  # type: Options
         self.source = self.executing.source  # type: Source
+        if hasattr(self.options, "skip_blank"):
```
[the remark this answers is not in the document]
Comment (aroberge) [reply stack_data/core.py:537]: Ok, for not checking that `self.options` has the correct type. If I use an enum, or something else with three options for `skip_blank`, I will have to do a test of some kind.

File: stack_data/core.py [hunk 6]
```diff
@@ -734,6 +751,16 @@ def lines(self) -> List[Union[Line, LineGap]]:
         for line in real_lines:
             line.leading_indent = leading_indent
 
+        if not self.skip_blank:
+            new_lines = [result[0]]
```
[the remark this answers is not in the document]
Comment (aroberge) [reply stack_data/core.py:755]: I was thinking of doing that, but I thought it would give a single piece of code with too convoluted logic. By separating it out, I thought it would be easier to understand. I'll try to fold it in. However, with the third case you want me to add, it might get confusing to a reader.

File: stack_data/formatting.py (hunk 2 above)
Comment (aroberge) [reply stack_data/formatting.py:141]: Ah, yes, good point.

**Comment by aroberge** [comment]
> You could add a Formatter keyword argument that requires a string which will be formatted in place and right-aligned to line up with the | from normal lines, so ": " (that's two spaces at the end) and ": |" would both be allowed and satisfy both your ideas. Some might also want to use a proper unicode vertical ellipsis with 3 dots.
Ok. Since this symbol would precede an empty line, there would b no need to add spaces after ":".

**Comment by alexmojaki** [comment]
btw `skip_blank` might not be a good name if there are three modes.

**Comment by aroberge** [comment]
While everything seems to work correctly, more improvements to the code needs to be done. I will try to finish it tomorrow.

**Comment by aroberge** [comment]
Other than missing unit tests, I believe that this is now ready. To illustrate the effect of each option, I will use the following example which includes many blank lines:

```
import stack_data
stack_data.Formatter().set_hook()

def last_item(seq):

    length = len(seq)


    return seq[

        length
    ]


seq = [1, 2, 3]
last = last_item(seq)
```

## Default output
The default is the current behaviour of `stack_data`, with a small cosmetic fix explained near the end of this comment.

```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 16, in <module>
       6 |     length = len(seq)
       9 |     return seq[
      10 |
      11 |         length
      12 |     ]
      15 | seq = [1, 2, 3]
-->   16 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 9, in last_item
       4 | def last_item(seq):
       6 |     length = len(seq)
-->    9 |     return seq[
                      ^^^^
      10 |
      11 |         length
               ^^^^^^^^^^
      12 |     ]
               ^
IndexError: list index out of range
```

## Making empty lines "visible" as text.

Setting the following parameters:

```
options = stack_data.Options(blank_lines=stack_data.BlankLines.VISIBLE)
stack_data.Formatter(options=options).set_hook()
```

The result is as follows:

```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 17, in <module>
       7 |     length = len(seq)
       8 |
       9 |
      10 |     return seq[
      11 |
      12 |         length
      13 |     ]
      14 |
      15 |
      16 | seq = [1, 2, 3]
-->   17 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 10, in last_item
       5 | def last_item(seq):
       6 |
       7 |     length = len(seq)
       8 |
       9 |
-->   10 |     return seq[
                      ^^^^
      11 |
      12 |         length
               ^^^^^^^^^^
      13 |     ]
               ^
IndexError: list index out of range
```

## Using only line number "markers" to indicate blank lines.

Setting the following parameters:

```
options = stack_data.Options(blank_lines=stack_data.BlankLines.LINE_NUMBER)
stack_data.Formatter(options=options).set_hook()
```

gives the following result (notice that `:` is used to indicate more than one blank line being skipped).

```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 17, in <module>
       7 |     length = len(seq)
       :
      10 |     return seq[
      11 |
      12 |         length
      13 |     ]
       :
      16 | seq = [1, 2, 3]
-->   17 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 10, in last_item
       5 | def last_item(seq):
       6 |
       7 |     length = len(seq)
       :
-->   10 |     return seq[
                      ^^^^
      11 |
      12 |         length
               ^^^^^^^^^^
      13 |     ]
               ^
IndexError: list index out of range
```

## Setting a different string for blank lines that are "skipped".

Using the following parameters

```
options = stack_data.Options(blank_lines=stack_data.BlankLines.LINE_NUMBER)
stack_data.Formatter(options=options, line_number_gap_string=": |").set_hook()
```

gives the following result:

```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 17, in <module>
       7 |     length = len(seq)
       : |
      10 |     return seq[
      11 |
      12 |         length
      13 |     ]
       : |
      16 | seq = [1, 2, 3]
-->   17 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 10, in last_item
       5 | def last_item(seq):
       6 |
       7 |     length = len(seq)
       : |
-->   10 |     return seq[
                      ^^^^
      11 |
      12 |         length
               ^^^^^^^^^^
      13 |     ]
               ^
IndexError: list index out of range
```

## Added explanation for a comment.

Using the default values, here's what happen if I were to remove the following code in `range_from_node`:

```
        if range_start == range_end == 0:
            # This is an empty line. If it were included, it would result
            # in a value of zero for the common indentation assigned to
            # a block of code.
            return None
```

Here is the result:

```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 16, in <module>
       6 |     length = len(seq)
       9 |     return seq[
      10 |
      11 |         length
      12 |     ]
      15 | seq = [1, 2, 3]
-->   16 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 9, in last_item
       4 | def last_item(seq):
       6 |     length = len(seq)
-->    9 |     return seq[
                      ^^^^
      10 |
      11 |         length
           ^^^^^^^^^^^^^^
      12 |     ]
           ^^^^^
IndexError: list index out of range
```
Notice how, due to the blank line 10 in the marked executing piece, the markers on lines 11 and 12 extend all the way back to the first column.

## Minor cosmetic bug fix

In `Formatter.format_line`, I now have the following:
```
                # if end <= start, we have an empty line inside a highlighted
                # block of code. In this case, we need to avoid inserting
                # an extra blank line with no markers present.
                if end > start:
                    result += (
                            " " * (start + len(prefix))
                            + self.executing_node_underline * (end - start)
                            + "\n"
                    )
```
If the `end <= start` test is removed, this is what happens:
```
Traceback (most recent call last):
 File "C:\Users\Andre\github\stack_data\example.py", line 16, in <module>
       6 |     length = len(seq)
       9 |     return seq[
      10 |
      11 |         length
      12 |     ]
      15 | seq = [1, 2, 3]
-->   16 | last = last_item(seq)
                  ^^^^^^^^^^^^^^
 File "C:\Users\Andre\github\stack_data\example.py", line 9, in last_item
       4 | def last_item(seq):
       6 |     length = len(seq)
-->    9 |     return seq[
                      ^^^^
      10 |

      11 |         length
               ^^^^^^^^^^
      12 |     ]
               ^
IndexError: list index out of range
```
Notice the gap between lines 10 and 11 in the marked piece.

File: stack_data/core.py [hunk 7]
```diff
@@ -387,6 +410,28 @@ def render(
         return ''.join(parts)
 
 
+class EmptyLine(Line):
```
Comment (alexmojaki) [inline stack_data/core.py:413]: Does this class need to exist? Is it just an optimisation of `render`?

File: stack_data/core.py [hunk 8]
```diff
@@ -709,6 +761,14 @@ def lines(self) -> List[Union[Line, LineGap]]:
...
             ):
                 result.append(LINE_GAP)
+            elif (prev_piece and add_empty_lines and piece.start > prev_piece.stop):
+                lines = []  # type: List[Union[EmptyLine, BlankLineRange]]
+                if self.blank_lines == BlankLines.LINE_NUMBER:
+                    lines.append(BlankLineRange(prev_piece.stop, piece.start-1))
+                else:  # BlankLines.VISIBLE
+                    for lineno in range(prev_piece.stop, piece.start):
+                        lines.append(EmptyLine(self, lineno))
+                result.extend(lines)
```
Comment (alexmojaki) [inline stack_data/core.py:771]:
```suggestion
            elif prev_piece and add_empty_lines and piece.start > prev_piece.stop:
                if self.blank_lines == BlankLines.LINE_NUMBER:
                    result.append(BlankLineRange(prev_piece.stop, piece.start-1))
                else:  # BlankLines.VISIBLE
                    for lineno in range(prev_piece.stop, piece.start):
                        result.append(EmptyLine(self, lineno))
```

`prev_piece` could be replaced by `pieces[i-1]` if you check that `i > 0`. But it probably wouldn't be an improvement.

File: stack_data/core.py [hunk 9]
```diff
@@ -46,6 +47,11 @@
 """
 
 
+class BlankLines(Enum):
+    HIDDEN = 1
+    VISIBLE = 2
+    LINE_NUMBER=3
```
Comment (alexmojaki) [inline stack_data/core.py:53]: This name doesn't feel like it explains itself well, especially since multiple blank lines specifically mean no line number gets shown. Maybe `COLLAPSED`?

**Review by alexmojaki — CHANGES_REQUESTED** [review]
Thanks, this is looking pretty good. My points are minor, the biggest remaining thing is the tests.

File: stack_data/core.py (hunk 7 above)
Comment (aroberge) [reply stack_data/core.py:413]: When I tried to simply use `Line`, an exception was raised. This was the minimal subclass that worked.

Comment (alexmojaki) [reply stack_data/core.py:413]: That's surprising, why didn't Line work?

File: stack_data/core.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aroberge) [reply stack_data/core.py]: Sorry, I clicked on resolved by accident and forgot about it.

File: stack_data/core.py (hunk 9 above)
Comment (aroberge) [reply stack_data/core.py:53]: I wasn't entirely satisfied with my choice of `LINE_NUMBER`, but it's the best I could come up with at the time.  I'll make the change to `COLLAPSED`.

File: stack_data/core.py (hunk 7 above)
Comment (aroberge) [reply stack_data/core.py:413]: Good question ... I just tried it, and it worked this time.

I'll make the other suggested changes quickly and will commit a new version.  

I'll investigate this, and start looking at adding unit tests after I finish work today.

File: stack_data/core.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aroberge) [reply stack_data/core.py]: This should now have been taken care of.

File: stack_data/core.py (hunk 9 above)
Comment (aroberge) [reply stack_data/core.py]: I've changed it to `SINGLE` with a docstring added describing what each name means.

File: stack_data/core.py (hunk 7 above)
Comment (aroberge) [reply stack_data/core.py]: I found out why using `Line` was raising an exception before; let's just say it was due to a user error.
I have not had time to add unit tests, but everything else should be done.

File: tests/golden_files/block_left_new.txt [hunk 10]
```diff
@@ -5,8 +5,9 @@ Traceback (most recent call last):
                                ^^^^^^^^^^
       73 |              for letter
                         ^^^^^^^^^^
-      74 |                in
+      74 |
```
Comment (aroberge) [inline tests/golden_files/block_left_new.txt:8]: This is actually a very difficult test to analyse visually. The content of this line ends with "74 | ". Notice the extra space at the end. When I was editing this "golden" text file with one editor that shall remain nameless, it would strip this ending space on that line, so that the test would fail.

**Comment by aroberge** [comment]
The code coverage has slightly decreased due to:

* The addition of a new `ValueError` exception that could be raised if `Options` is not initialized correctly.
* The absence of testing for the `BlankLines.SINGLE` when an exception is not raised.

Are unit tests needed for these two cases?

File: stack_data/formatting.py [hunk 11]
```diff
@@ -168,7 +179,7 @@ def format_line(self, line: Line) -> str:
             result += " "
 
         if self.show_linenos:
-            result += "{:4} | ".format(line.lineno)
+            result += self.line_number_format_string.format(line.lineno)
```
Comment (alexmojaki) [inline stack_data/formatting.py:182]: Why this new option? It seems unrelated to the rest of the PR. And shouldn't it be used in `format_blank_lines_linenumbers`?

File: stack_data/formatting.py [hunk 12]
```diff
@@ -184,14 +195,28 @@ def format_line(self, line: Line) -> str:
...
+                            + "\n"
+                    )
         return result
 
+
+    def format_blank_lines_linenumbers(self, blank_line):
+        if self.current_line_indicator:
+            result = " " * len(self.current_line_indicator)
+        else:
+            result = "   "
```
Comment (alexmojaki) [inline stack_data/formatting.py:214]: https://coveralls.io/builds/51814297/source?filename=stack_data%2Fformatting.py#L214 points out that this line is uncovered, and it seems worth testing.

**Comment by alexmojaki** [comment]
> The code coverage has slightly decreased due to:
> 
> * The addition of a new `ValueError` exception that could be raised if `Options` is not initialized correctly.
> * The absence of testing for the `BlankLines.SINGLE` when an exception is not raised.
> 
> Are unit tests needed for these two cases?

The first point seems trivial to test. Just construct a bad formatter and assert the raised exception. No need to do any formatting.

I don't know what the second point means.

**Review by alexmojaki — CHANGES_REQUESTED** [review]

File: stack_data/formatting.py (hunk 11 above)
Comment (aroberge) [reply stack_data/formatting.py:182]: I added it as the indentation was fixed and I might want a larger value in friendly_traceback; it is the only hard-coded value you had in stack_data, while all the others could be changed by the users. And, yes, I overlooked the fact that it should have been used in `format_blank_lines_linenumbers` as well. I will need to do this.

File: stack_data/formatting.py (hunk 12 above)
Comment (aroberge) [reply stack_data/formatting.py:214]: Ok. I probably won't have time to do this tonight though.

**Comment by aroberge** [comment]
> > * The absence of testing for the `BlankLines.SINGLE` when an exception is not raised.
> > 
> > Are unit tests needed for these two cases?
> 
> The first point seems trivial to test. Just construct a bad formatter and assert the raised exception. No need to do any formatting.
> 
> I don't know what the second point means.
I'm thinking of something like `print_stack` when many blank lines are printed. I didn't try it out to confirm that it would show changes.  I'll need to do some explicit tests to confirm that a unit test might be needed.

File: stack_data/formatting.py (hunk 12 above)
Comment (aroberge) [reply stack_data/formatting.py]: I had added this `else` part based on the presence of the second term in https://github.com/alexmojaki/stack_data/blob/7e8808252fde3fbd5e5fbd84ae2913e98cc77fde/stack_data/formatting.py#L173
However, it turns out that `result` was never null and that this line was redundant in the original code, based on all existing tests. So I removed the corresponding `else` part of my addition.

Comment (alexmojaki) [reply stack_data/formatting.py]: It's pretty easy to see when the line isn't redundant in the previous code: when you set `show_linenos=False, current_line_indicator=""`. `result` won't be null, it'll be an empty string. Defaulting it to spaces ensures that the user's code is indented like in a standard traceback. I guess it should have been tested before, but coverage didn't see the gap.

Comment (aroberge) [reply stack_data/formatting.py]: Do you want me to reinstate it and write a relevant test for it?

Comment (alexmojaki) [reply stack_data/formatting.py]: Yes please.

File: tests/test_formatter.py [hunk 13]
```diff
@@ -94,3 +94,31 @@ def check_example(name):
...
+            MyFormatter(show_linenos=False, options=Options(blank_lines=BlankLines.VISIBLE)).print_exception()
+
+
+def test_invalid_single_option():
+    try:
+        MyFormatter(show_linenos=False, options=Options(blank_lines=BlankLines.SINGLE))
+    except ValueError:
+        assert True
+    else:
+        assert False
```
Comment (alexmojaki) [inline tests/test_formatter.py:129]:
```suggestion
    with pytest.raises(ValueError):
        MyFormatter(show_linenos=False, options=Options(blank_lines=BlankLines.SINGLE))
```

and an `import pytest` at the top.

File: stack_data/formatting.py (hunk 12 above)
Comment (alexmojaki) [reply stack_data/formatting.py]: Looks like there's been some confusion or misremembering. The test needs to specifically have both the arguments `show_linenos=False, current_line_indicator=""` in one formatter.

File: tests/test_formatter.py (hunk 13 above)
Comment (aroberge) [reply tests/test_formatter.py:129]: Will do.

File: stack_data/formatting.py (hunk 12 above)
Comment (aroberge) [reply stack_data/formatting.py]: Yes, I was wondering why the line `result = result or "   "` did not seem to have any effect. I will fix this.

File: stack_data/formatting.py [hunk 14]
```diff
@@ -166,12 +177,13 @@ def format_line(self, line: Line) -> str:
             else:
                 result = " " * len(self.current_line_indicator)
             result += " "
+        else:
```
Comment (alexmojaki) [inline stack_data/formatting.py:180]: I think this should be removed, the original line `result = result or " "` is enough. A test without `current_line_indicator` but with `show_linenos` would confirm.

Comment (aroberge) [reply stack_data/formatting.py:180]: You are, of course, correct. I had a brain freeze when I added this.

Comment (aroberge) [reply stack_data/formatting.py:180]: I've tested this by creating two additional tests (with or without blank lines included). Do you want these extra tests added in a commit?

Comment (alexmojaki) [reply stack_data/formatting.py:180]: If you've already written the test etc. then might as well. But no need for extra effort. Just remove these two lines, unless the tests show otherwise.

Comment (aroberge) [reply stack_data/formatting.py:180]: With the two lines, the result is as follows:
```
Traceback (most recent call last):
 File "formatter_example.py", line 85, in blank_lines
     79 | def blank_lines():
     80 |     a = [1, 2, 3]
     82 |     length = len(a)
     85 |     return a[length]
                     ^^^^^^^^^
IndexError: list index out of range
```
Without them, the indentation is slightly less
```
Traceback (most recent call last):
 File "formatter_example.py", line 85, in blank_lines
  79 | def blank_lines():
  80 |     a = [1, 2, 3]
  82 |     length = len(a)
  85 |     return a[length]
                  ^^^^^^^^^
IndexError: list index out of range
```
Which one do you prefer?  Once decided, it probably would make sense to include the tests.

Comment (alexmojaki) [reply stack_data/formatting.py:180]: Oh, it looks nicer with the extra spaces, I didn't expect that. Especially if the line numbers reach 4 digits. So let's go with the former option, and then I think `result = result or " "` isn't needed.

Comment (aroberge) [reply stack_data/formatting.py:180]: I needed to modified the code when using `blank_lines.HIDDEN` with no current line marker so that it would be consistent with the others. I added three tests, one for each setting.

**Review by alexmojaki — APPROVED** [review]
All looks good, thanks! Happy for me to merge?

**Comment by aroberge** [comment]
Yes, please do!
